kissim/comparison/tree.py

```diff
@@ -0,0 +1,309 @@
+    tree : scipy.cluster.hierarchy.ClusterNode
+        Clustering tree.
+    distance_matrix : pandas.DataFrame
+        Distance matrix on which clustering is based.
+    node_means : dict of int: float
+        Mean distance (value) for each node index (key).
+    similarity : bool
+        If `True`, convert distance matrix into similarity matrix using
+        `1 - distance_matrix / distance_matrix.max().max()`, else use distance matrix directly.
+
+    Returns
+    -------
+    node_means : dict of int: float
+        Mean similarities (value) for each node index (key).
+        Note: The return value is the populated input `node_means` object.
+    """
+
+    # Calculate mean similarities or mean distances?
+    if similarity:
+        matrix = (1 - distance_matrix / distance_matrix.max().max()).copy()
+        leaf_mean = 1
+    else:
+        matrix = distance_matrix.copy()
+        leaf_mean = 0
+
+    # Set mean for root node
+    if tree.id not in node_means:
+        # Get full matrix (root includes full dataset)
+        matrix_all_clusters = matrix.values
+        # Get matrix values from lower triangle without diagonal
+        values_all_clusters = matrix_all_clusters[
+            np.tril_indices(matrix_all_clusters.shape[0], -1)
+        ]
+        # Calculate mean of full dataset
+        node_means[tree.id] = np.average(values_all_clusters)
+
+    # Get mean only for nodes that are not a leaf
+    if not tree.is_leaf():
+
+        # Iterate over the nodes to the right and the left of the current node
+        for left in [0, 1]:
+            if left:
+                # Get list of indices cluster left (A)
+                indices = tree.get_left().pre_order(lambda x: x.id)
+            else:
+                # Get list of indices cluster right (B)
+                indices = tree.get_right().pre_order(lambda x: x.id)
+
+            if len(indices) > 1:
+                # Get matrix subset for cluster members only
+                matrix_by_cluster = matrix.iloc[indices, indices].values
+                # Get matrix values from lower triangle without diagonal
+                values_by_cluster = matrix_by_cluster[
+                    np.tril_indices(matrix_by_cluster.shape[0], -1)
+                ]
+                # Calculate mean within a cluster
+                if left:
+                    node_means[tree.get_left().id] = np.average(values_by_cluster)
+                    _get_node_means(tree.get_left(), distance_matrix, node_means, similarity)
+                else:
+                    node_means[tree.get_right().id] = np.average(values_by_cluster)
+                    _get_node_means(tree.get_right(), distance_matrix, node_means, similarity)
+            # Set leaf "mean" to identity (similarity=1; distance=0)
+            else:
+                if left:
+                    node_means[tree.get_left().id] = leaf_mean
+                else:
+                    node_means[tree.get_right().id] = leaf_mean
+
```

Comment:
In case of the leaves, I also added for both the right and left leaf "1"/identity in case of similarity, this was set only for the [left leaf](https://github.com/AJK-dev/kissim/blob/926aa5291b7f768f85c27bc6cb976c4db808401f/kissim/comparison/kissim_to_newick.py#L86).

Again, happy to delete this if it does not make sense; no effect on the newick format as far as I understand.

Reply:
Good that you caught this. Indeed good to keep it this way as it might be useful in other circumstances/future formats.